### PR TITLE
cmor tool improvements

### PR DESCRIPTION
**Pull request by ilaflott**
- [x] I ran my code
- [x] I tried to make my code readable
- [x] I tried to comment my code
- [x] I wrote a new test, if applicable
- [x] I wrote new instructions/documentation, if applicable
- [x] I ran pytest and inspected it's output
- [x] I ran pylint and attempted to implement some of it's feedback

This PR...
- overhauls the `README.md` with `fre-cli` oriented usage and removes outdated doc
- refactors `cmor_mixer.py` heavily to be more readable, have more modularity to it to clearly illustrate who-does-what and why, with an emphasis on clearly exposing data movement inherent in the routine
- functions now have doc strings with argument descriptions in most places
- better run-time feedback of functioning/IO/logic etc during runtime, hopefully not too much
- better comments where possible
- new test case where `key:value` pair in `var_list` argument differ
- more test case cleanup/setup, more specific test case success conditions
- new CLI/`click.testing.CliRunner` test for the new functionality of the `fre.cmor` module
- change of the way the click module interfaces with the cmor run subtool module

### CODE REVIEW

File: .github/workflows/create_test_conda_env.yml [hunk 1]
```diff
@@ -59,8 +59,8 @@ jobs:
           python --version
           $CONDA/envs/fre-cli/bin/python --version
           
-          # run pylint
-          pylint -ry --ignored-modules netCDF4 fre/ || echo "pylint returned non-zero exit code. preventing workflow from dying with this echo."
+          # run pylint, ignored modules avoid warnings arising from code internal to those modules 
+          pylint --max-args 6 -ry --ignored-modules netCDF4,cmor fre/ || echo "pylint returned non-zero exit code. preventing workflow from dying with this echo."
```
Comment (ilaflott) [inline .github/workflows/create_test_conda_env.yml:63]: 5 arguments always seems like so few... `pylint` will criticize `cmor` and blame your code, hence the new `ignored-modules` adjustment

File: fre/cmor/README.md [hunk 2]
```diff
@@ -1,83 +1,95 @@
...
 
+new readme instructions iwll replace these. 
+
+
+```
 # Before start this script in common way run these 2 command in terminal where you are going to execute this script:
-#     module load python/3.9
-#     conda activate cmor
+>     module load python/3.9
+>     conda activate cmor
```
Comment (ilaflott) [inline fre/cmor/README.md:9]: remove

File: fre/cmor/README.md [hunk 3]
```diff
@@ -1,83 +1,95 @@
...
+> ls cmip6-cmor-tables/Tables
+...
+    CMIP6_CV.json
+    CMIP6_formula_terms.json
+    CMIP6_grids.json
+    CMIP6_coordinate.json
+...
+```
+
+Detailed description of program is placed at https://docs.google.com/document/d/1HPetcUyrVXDwCBIyWheZ_2JzOz7ZHi1y3vmIlcErYeA/edit?pli=1
```
Comment (ilaflott) [inline fre/cmor/README.md:65]: remove

File: fre/cmor/frecmor.py [hunk 4]
```diff
@@ -33,7 +33,23 @@ def cmor_cli():
...
+    )
+    #    context.forward(
+    #        _cmor_run_subtool(
+    #            indir = indir,
+    #            json_var_list = varlist,
+    #            json_table_config = table_config,
+    #            json_exp_config = exp_config,
+    #            outdir = outdir
+    #        )
+    #    )
```
Comment (ilaflott) [inline fre/cmor/frecmor.py:52]: clenaup

File: fre/tests/test_files/varlist_local_target_vars_differ [hunk 5]
```diff
@@ -0,0 +1,3 @@
+{
+    "sosV2": "sos"
```
Comment (ilaflott) [inline fre/tests/test_files/varlist_local_target_vars_differ:2]: glad i figured out how to make this work. there was some serious confusion before in the code regarding the way the key and value get used respectively

File: fre/cmor/cmor_mixer.py [hunk 6]
```diff
@@ -1,408 +1,622 @@
 #!/usr/bin/env python
```
Comment (ilaflott) [inline fre/cmor/cmor_mixer.py:1]: i'm tempted to remove this because it could unintentionally point someone to a python where things *kinda* work, but don't actually. it hardly seems necessary for functioning anyways.

either it points to a `python2` nearby, and things should error quickly. Or it points to a `python3` nearby with just enough bells and whistles to not fail immediately and clearly

File: fre/cmor/cmor_mixer.py [hunk 7]
```diff
@@ -1,408 +1,622 @@
...
 import cmor
 
 
+# ----- \start consts
+
+# ----- \end consts
+
+################################
+# ----- SMALLER ROUTINES ----- #
+################################
```
Comment (ilaflott) [inline fre/cmor/cmor_mixer.py:22]: cleanup

File: fre/cmor/cmor_mixer.py [hunk 8]
```diff
@@ -1,408 +1,622 @@
...
-    print(f"(netcdf_var) cmip_input_json = {cmip_input_json}")
-    print(f"(netcdf_var) cmor_table_vars_file = {cmor_table_vars_file}")
 
-    # load variable list (CMOR table)
-    cmor.load_table(cmor_table_vars_file)
-    var_list = list(ds.variables.keys())
-    print(f"(netcdf_var) list of variables: {var_list}")
+#############################
+# ----- BULK ROUTINES ----- #
+#############################
```
Comment (ilaflott) [inline fre/cmor/cmor_mixer.py:165]: cleanup

File: fre/cmor/cmor_mixer.py [hunk 9]
```diff
@@ -1,408 +1,622 @@
...
-    # read the input units
-    var = ds[gfdl_var][:]
-    var_dim = len(var.shape)
+def rewrite_netcdf_file_var ( proj_table_vars = None,
+                              local_var = None,
+                              netcdf_file = None,
+                              target_var = None,
+                              json_exp_config = None,
+                              json_table_config = None,
+                              tmp_dir = None
```
Comment (ilaflott) [inline fre/cmor/cmor_mixer.py:174]: not all of these need their own line

File: fre/cmor/cmor_mixer.py [hunk 10]
```diff
@@ -1,408 +1,622 @@
...
+
+    print(f"(rewrite_netcdf_file_var) var_dim = {var_dim}, local_var = {local_var}")
+
+
+
+    # now we set up the cmor module object
+    # initialize CMOR
+    cmor.setup(
+        netcdf_file_action    = cmor.CMOR_PRESERVE,
+        set_verbosity         = cmor.CMOR_QUIET, #default is CMOR_NORMAL
```
Comment (ilaflott) [inline fre/cmor/cmor_mixer.py:252]: opinions on `CMOR_QUIET`? noisy module.

File: fre/cmor/cmor_mixer.py [hunk 11]
```diff
@@ -1,408 +1,622 @@
...
-    # read the positive attribute
-    var = ds[gfdl_var][:]
-    positive = proj_table_vars["variable_entry"] [var_j] ["positive"]
-    print(f"(netcdf_var) var_lst[{gfdl_var}] = {var_j}, positive = {positive}")
+    # read positive attribute and create cmor_var?
+    positive = proj_table_vars["variable_entry"] [target_var] ["positive"]
+    print(f"(rewrite_netcdf_file_var) positive = {positive}")
+    ##cmor_var = cmor.variable(local_var, units, axes)
+    #cmor_var = cmor.variable(local_var, units, axes, positive = positive)
+    #cmor_var = cmor.variable(target_var, units, axes)
```
Comment (ilaflott) [inline fre/cmor/cmor_mixer.py:360]: cleanup

File: fre/cmor/cmor_mixer.py [hunk 12]
```diff
@@ -1,408 +1,622 @@
...
-    ''' primary steering function for the cmor_mixer tool, i.e
-    essentially main '''
-    print('\n\n----- START _cmor_run_subtool call -----')
-
+            print(f"(cmorize_target_var_files) mv_cmd = {mv_cmd}")
+            #os.system(mv_cmd)
+            subprocess.run(mv_cmd, shell=True, check=True)
+
+
+        # delete files in work dirs
```
Comment (ilaflott) [inline fre/cmor/cmor_mixer.py:497]: cleanup

File: fre/cmor/cmor_mixer.py [hunk 13]
```diff
@@ -1,408 +1,622 @@
...
+            local_var, # OK
+            iso_datetime_arr, # OK
+            name_of_set, # OK
+
+            json_exp_config, # this makes sense just fine, i think
+            outdir, # OK
+
+            proj_table_vars, # passing this INSTEAD OF json_table_config makes more sense to me
+            json_table_config, # if this is being passed why pass proj table vars?
+            # --> b.c. the cmor module reads it directly!
```
Comment (ilaflott) [inline fre/cmor/cmor_mixer.py:607]: cleanup

File: fre/cmor/tests/test_cmor_run_subtool.py [hunk 14]
```diff
@@ -80,9 +126,124 @@ def test_fre_cmor_run_output_compare(capfd):
...
+    # VERY ANNOYING !!! FYI WARNING TODO
+    if Path(OUTDIR).exists():
+        try:
+            shutil.rmtree(OUTDIR)
+        except OSError as exc:
+            print(f'WARNING: OUTDIR={OUTDIR} could not be removed.')
+            print( '         this does not matter that much, but is unfortunate.')
+            print( '         supicion: something the cmor module is using is not being closed')
+
+    #assert not Path(OUTDIR).exists()    # VERY ANNOYING !!! FYI WARNING TODO
```
Comment (ilaflott) [inline fre/cmor/tests/test_cmor_run_subtool.py:173]: these cleanup steps throw some very annoying error messages if you don't let the exception go-

```
...
...
...
fre/cmor/tests/test_cmor_run_subtool.py::test_fre_cmor_run_subtool_case1_output_compare_metadata PASSED                                                                            [ 55%]
fre/cmor/tests/test_cmor_run_subtool.py::test_setup_fre_cmor_run_subtool_case2 FAILED                                                                                              [ 66%]

======================================================================================== FAILURES ========================================================================================
_________________________________________________________________________ test_setup_fre_cmor_run_subtool_case2 __________________________________________________________________________

capfd = <_pytest.capture.CaptureFixture object at 0x2b0f0acad790>

    def test_setup_fre_cmor_run_subtool_case2(capfd):
        ''' make a copy of the input file to the slightly different name.
        checks for outputfile from prev pytest runs, removes it if it's present.
        this routine also checks to make sure the desired input file is present'''
        if Path(FULL_OUTPUTFILE).exists():
            Path(FULL_OUTPUTFILE).unlink()
        assert not Path(FULL_OUTPUTFILE).exists()
    
        if Path(OUTDIR+'/CMIP6').exists():
            shutil.rmtree(OUTDIR+'/CMIP6')
        assert not Path(OUTDIR+'/CMIP6').exists()
    
    
        # VERY ANNOYING !!! FYI WARNING TODO
        if Path(TMPDIR).exists():
            #try:
>           shutil.rmtree(TMPDIR)

fre/cmor/tests/test_cmor_run_subtool.py:156: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../conda/envs/fre_cli/lib/python3.9/shutil.py:734: in rmtree
    _rmtree_safe_fd(fd, path, onerror)
../../conda/envs/fre_cli/lib/python3.9/shutil.py:667: in _rmtree_safe_fd
    _rmtree_safe_fd(dirfd, fullname, onerror)
...
...
...
../../conda/envs/fre_cli/lib/python3.9/shutil.py:690: in _rmtree_safe_fd
    onerror(os.unlink, fullname, sys.exc_info())
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

topfd = 28, path = 'fre/tests/test_files/outdir/tmp/CMIP6/CMIP6/ISMIP6/PCMDI/PCMDI-test-1-0/piControl-withism/r3i1p1f1/Omon/sos/gn/v20241022'
onerror = <function rmtree.<locals>.onerror at 0x2b0f0accbe50>

    def _rmtree_safe_fd(topfd, path, onerror):
...
...
...
...
>                   os.unlink(entry.name, dir_fd=topfd)
E                   OSError: [Errno 16] Device or resource busy: '.nfs00000000a9f8735600000090'

../../conda/envs/fre_cli/lib/python3.9/shutil.py:688: OSError
================================================================================ short test summary info =================================================================================
FAILED fre/cmor/tests/test_cmor_run_subtool.py::test_setup_fre_cmor_run_subtool_case2 - OSError: [Errno 16] Device or resource busy: '.nfs00000000a9f8735600000090'
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! stopping after 1 failures !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
============================================================================== 1 failed, 5 passed in 2.65s ===============================================================================
```

File: fre/cmor/cmor_mixer.py [hunk 15]
```diff
@@ -1,408 +1,622 @@
...
+        os.chdir(gotta_go_back_here+tmp_dir) # this is unavoidable, cmor module FORCES write to CWD
+        local_file_name = rewrite_netcdf_file_var( proj_table_vars                       ,
+                                                   local_var                             ,
+                                                   gotta_go_back_here + nc_file_work     ,
+                                                   target_var                            ,
+                                                   gotta_go_back_here + json_exp_config  ,
+                                                   gotta_go_back_here + json_table_config,
+                                                   gotta_go_back_here + tmp_dir            )
+        os.chdir(gotta_go_back_here)
+        assert Path( gotta_go_back_here+tmp_dir+local_file_name ).exists()
```
Comment (ilaflott) [inline fre/cmor/cmor_mixer.py:459]: @cwhitlock-NOAA do you see what `cmor` made me do?

**Comment by ilaflott** [comment]
this will close #204

**Comment by ilaflott** [comment]
[come a long way havent we...](https://github.com/NOAA-GFDL/fre-cli/tree/edea8f44e97d488aa6a50944d430c33dac39e31c/fre/cmor)

**Comment by ilaflott** [comment]
regarding meniton in https://github.com/NOAA-GFDL/CatalogBuilder/pull/79 

links in commit point to https://github.com/NOAA-GFDL/MDTF-diagnostics/tree/main/data/gfdl-cmor-tables

note submodule dep: https://github.com/NOAA-GFDL/MDTF-diagnostics/tree/main/data/cmip6-cmor-tables comes from https://github.com/PCMDI/cmip6-cmor-tables

File: fre/cmor/tests/test_cmor_run_subtool.py (hunk 14 above)
Comment (ilaflott) [reply fre/cmor/tests/test_cmor_run_subtool.py:173]: leaving this unresolved for now. the directory re-creation for the automatic test pipeline is a nice-to-have, not essential. as if you remove the directory by hand after erroring, and restart at this failed test, the test succeeds as expected. 

largely that re-tests the directory creation that already succeeded in the prev test, so its even a tad redundant.

Comment (ilaflott) [reply fre/cmor/tests/test_cmor_run_subtool.py:173]: another reason to leave it unresolved- i suspect this is `cmor`'s fault...

File: fre/cmor/cmor_mixer.py [hunk 16]
```diff
@@ -1,408 +1,596 @@
...
+    print(f"(cmorize_target_var_files) outdir = {outdir}")
+    tmp_dir = None
+    if any( [ outdir == "/local2",
+              outdir.find("/work") != -1,
+              outdir.find("/net" ) != -1 ] ):
+        print(f'(cmorize_target_var_files) using /local /work /net ( tmp_dir = {outdir}/ )')
+        tmp_dir = "{outdir}/"
+    else:
+        print('(cmorize_target_var_files) NOT using /local /work /net (tmp_dir = outdir/tmp/ )')
+        tmp_dir = f"{outdir}/tmp/"
```
Comment (ceblanton) [inline fre/cmor/cmor_mixer.py:153]: I believe you that this is needed for some reason, but it's odd. This is the logic, right?

If the output directory is on /work or /net2, then set tmpdir to that directory.

If the output directory is somewhere else, set tmpdir to the current directory.

File: fre/cmor/cmor_mixer.py [hunk 17]
```diff
@@ -1,408 +1,596 @@
...
+    # read in time_bnds , if present
     time_bnds = []
-    print(f"(netcdf_var) time_coord_units = {time_coord_units}")
-    print(f"(netcdf_var) time_bnds  = {time_bnds}")
     try:
-        print( f"(netcdf_var) Executing cmor.axis('time', \n"
-               f"(netcdf_var) coord_vals = \n{time_coords}, \n"
-               f"(netcdf_var) cell_bounds = {time_bnds}, units = {time_coord_units})   " )
-
         time_bnds = ds["time_bnds"][:]
```
Comment (ceblanton) [inline fre/cmor/cmor_mixer.py:206]: This is a non-change, but just noticed it. It's not a problem but I'll point it out since you like netcdf/cf standards now :)

Recall how you are grabbing the dimension names with the CF standards `axis`?

time bounds is actually the same, but by convention it's usually called "time_bnds". The CF rule is that the "time" variable should have a "bounds" attribute pointing to the name of the time bounds variable.

File: fre/cmor/cmor_mixer.py [hunk 18]
```diff
@@ -1,408 +1,596 @@
...
+            print( "(cmor_run_subtool)         ... target_var not found in CMOR variable group")
+            continue
+
+        # it is in there, get the name of the data inside the netcdf file.
+        target_var=var_list[local_var] # often equiv to local_var but not necessarily.
+        if local_var != target_var:
+            print(f'(cmor_run_subtool) WARNING: local_var == {local_var} '
+                  f'!= {target_var} == target_var')
+            print(f'i am expecting {local_var} to be in the filename, and i expect the variable'
+                  f' in that file to be {target_var}')
```
Comment (ceblanton) [inline fre/cmor/cmor_mixer.py:576]: I think I'm following but disagree with the message here. Shouldn't it be:

I am expecting {local_var} to be in the filename, and I expect the variable in that file to be {local_var}. The output file will be named {target_var}.

**Review by ceblanton — APPROVED** [review]
Your code deserved a better review than what I did. I didn't look hard at the testing additions, but the rest looked great with the few comments I had. I don't think any of the comments are deal-breakers tho for taking this.

File: fre/cmor/cmor_mixer.py [hunk 19]
```diff
@@ -1,408 +1,596 @@
...
 
-    # output file
-    dsout = nc.Dataset(out_nc,
-                       "w") #, format = "NETCDF3_CLASSIC")
+    # output file, same exact data_model as input file.
+    # note- totally infuriating...
+    #       the correct value for the format arg is netCDF4.Dataset.data_model
+    #       and NOT netCDF4.Dataset.disk_format
+    dsout = nc.Dataset(out_nc, "w",
+                       format = dsin.data_model)
```
[the remark this answers is not in the document]
Comment (ilaflott) [reply fre/cmor/cmor_mixer.py:40]: i wanted to respect the fileformat as a default choice for now! hard to understand the original scripts full intentions with the commented-out netcdf3 classic specification...

is output format specification desired? such that a specific output target format can be written out? I think the `cmor` module may enforce netcdf4 anywho...

File: fre/cmor/cmor_mixer.py (hunk 16 above)
Comment (ilaflott) [reply fre/cmor/cmor_mixer.py:153]: I am not entirely convinced this is needed for any reason. If there exists one, it's clear to someone else, not me. I only attempted only to preserve this directory creation logic as I originally encountered it.

Your read of what's happening isn't exaaaaactly right, to split hairs...
- if the output directory is '/local2', or it's on '/work', or '/net', then use the output directory specified
- otherwise, create a new directory within outdir called 'tmp' and use that.

TLDR- We can put in whatever file movement logic here we want! I am not entirely sure whats best for gfdl/ppan, but if someone can tell me, i'll implement it.

File: fre/cmor/cmor_mixer.py (hunk 17 above)
Comment (ilaflott) [reply fre/cmor/cmor_mixer.py:206]: ahhhhh, I see. That's a good idea. 

If you don't mind however, this kind of checking for bounds attributes is going to come up when i'm checking other coordinate/dimension things in PR #218. There, I'm working on a few different conditional-checks to read data in a bit more dynamically.

File: fre/cmor/cmor_mixer.py [hunk 20]
```diff
@@ -1,408 +1,596 @@
...
+### ------ helper functions  ------ ###
 def copy_nc(in_nc, out_nc):
-    ''' copy a net-cdf file, presumably '''
-    print('\n\n----- START copy_nc call -----')
+    '''
+    copy target input netcdf file in_nc to target out_nc. I have to think this is not a trivial copy
+    operation, as if it were, using shutil's copy would be sufficient. accepts two arguments
+        in_nc: string, path to an input netcdf file we wish to copy
+        out_nc: string, an output path to copy the targeted input netcdf file to
+    '''
```
[the remark this answers is not in the document]
Comment (ilaflott) [reply fre/cmor/cmor_mixer.py:28]: I don't see any EXPLICIT need, no. i imagine this may be batch-job oriented movement, perhaps?

File: fre/cmor/cmor_mixer.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (ceblanton) [reply fre/cmor/cmor_mixer.py:40]: Respecting the input format is usually the right thing to do, but for CMOR it might be worth only supporting Netcdf-4 with mild compression (as we did for CMIP6).

File: fre/cmor/cmor_mixer.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (ceblanton) [reply fre/cmor/cmor_mixer.py:28]: well, let's just keep it in mind as something maybe-pathological that we can adjust later

File: fre/cmor/cmor_mixer.py (hunk 16 above)
Comment (ceblanton) [reply fre/cmor/cmor_mixer.py:153]: Alright, I see.  Next time we are confused by the different tmpdir behavior tho let's make them the same. It looks like they are really the same anyway, with a confusing twist on the /local2 /work thing that doesn't even matter.

**Comment by ilaflott** [comment]
@ceblanton has good feedback here, most of which is actually not straightforward to address, or involves some creative decision making. given that this PR is gigantic but functional, i'm inclined to push that feedback forward for now, and spin those up into issues and/or dev efforts on the next PR #218
